src/main/java/com/cgm/infolab/db/repository/ChatMessageRepository.java

```diff
@@ -90,4 +90,44 @@ public List<ChatMessageEntity> getByRoomName(String roomName) {
             return new ArrayList<>();
         }
     }
+
+    public List<ChatMessageEntity> getByRoomNameNumberOfMessages(String roomName, int numberOfMessages) {
+        RoomEntity room = roomRepository.getByRoomName(roomName).orElseThrow(() -> {
```

Comment:
Ok l'aggiunta del nuovo metodo, ma dobbiamo togliere un po' di duplicazione introdotta coi copia/incolla.

Una prima duplicazione è molto evidente, e anche di facile risoluzione. Parlo della lambda che trasforma il resul set del Db nell'entity ChatMessage. Estraiamo un metodo privato che prende in input `ResultSet rs, int rowNum` e poi passiamo la reference del metodo privato con `this::mioMetodoMapper`. Un metodo che fa questa trasformazione è detto "mapper".

L'altra duplicazione è meno evidente ma comunque relativamente semplice da rimuovere. Possiamo osservare che ciò che effettivamente cambia nei due metodi è solo la query (e i suoi parametri). Anche qui estrarrei un metodo private che poi farei richiamare da quelli pubblici. Il metodo che prende il numero di messaggi come parametro, mantiene il controllo sul numero positivo, e se negativo, richiama l'altro metodo pubblico senza parametri

Reply:
Potresti incontrare una difficoltà, che ho volutamente evitato di spiegarti 😝. Però ti lascio un indizio https://stackoverflow.com/questions/2925153/can-i-pass-an-array-as-arguments-to-a-method-with-variable-arguments-in-java

Reply:
Non so se ho fatto come intendevi. In caso fammi sapere che correggo : )

Reply:
Quasi 😅. Partiamo dall'osservazione che
```java
RoomEntity room = roomRepository.getByRoomName(roomName).orElseThrow(() -> {
            throw new IllegalArgumentException(String.format("Room roomName=\"%s\" non trovata.", roomName));
        });
```
è duplicato a sua volta e quindi abbiamo il suggerimento che in qualche modo sia da mettere a fattor comune. La [duplicazione](https://refactoring.guru/smells/duplicate-code) è un **code smell** e in quanto tale è un indicatore di un probabile difetto di design. 

Una prima semplice azione, che non rimuove la duplicazione ma che ne riduce l'impatto, sarebbe fare anche di questo pezzo un metodo privato e richiamarlo al posto del codice duplicato. Non ci risoverebbe il problema, tuttavia lo ridurremo di molto, e quindi sarebbe già un primo passo da fare. Diciamo che è meglio farlo che non farlo.

Per risolvere del tutto il problema dobbiamo invece prendere quel pezzo e metterlo nel nuovo metodo privato `get` (get non è un buon nome per questo metodo, lo chiamerei tipo `getRoomsIfCondition`). Anche se non più duplicato, rimane comunque una buona idea farne un metodo privato che abbia un nome esplicativo dell'intento.

A questo punto la duplicazione è rimossa. Però il metodo `get` è un po' troppo complesso. Ci sono un po' troppe `if` il chè lo rende di difficile manutenzione ([complessità ciclomatica](https://www.google.com/search?q=cyclomatic+complexity&oq=cyclom&aqs=edge.1.69i57j0i512l8.10568j0j1&sourceid=chrome&ie=UTF-8)). La manutenzione non riguarda tanto la forma che il metodo ha adesso, ma quella che potrebbe assumere nel futuro. Pensa ad esempio se ti chiedessi di aggiungere una terza clausola. Poi una quarta, quinta e via dicendo. Ci sono anche qui almeno un paio di maniere di abbassare la complessità. Una con la if e una senza (in ordine di difficoltà, [difficile vs complesso](https://www.google.com/search?q=difficult+vs+complex+programming+talk&sxsrf=AJOqlzWhiQQLSXJbINdgMfNeV11T8OJRSQ%3A1677061591496&ei=1-31Y9v4HdnXxc8Pgbu74AY&ved=0ahUKEwibwLWj9aj9AhXZa_EDHYHdDmwQ4dUDCA8&uact=5&oq=difficult+vs+complex+programming+talk&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQAzIECCEQFTIECCEQFToKCAAQRxDWBBCwAzoLCAAQFhAeEPEEEAo6CQgAEB4QDRDxBDoLCAAQHhANEA8Q8QQ6CAgAEBYQHhAPOgkIABAWEB4Q8QQ6CggAEAUQHhANEA86DQgAEAUQHhANEA8Q8QQ6BwghEKABEAo6BQghEKABSgQIQRgAUIEKWMYrYPctaANwAXgAgAGEAYgB9QiSAQM5LjOYAQCgAQHIAQXAAQE&sclient=gws-wiz-serp), **Complicated implies being difficult to understand but with time and effort, ultimately knowable. Complex, on the other hand, describes the interactions between a number of entities.**)
E' difficile da spiegare... ti faccio un ramo con le due soluzioni su due commit differenti, poi te lo link in un commento qui sotto

Reply:
Guarda questo ramo [BE/rimozione-duplicazione](https://github.com/Studiofarma/Infolab/commits/BE/rimozione-duplicazione). Guarda gli ultimi 6 commit uno alla volta e osserva i singoli passi di refactoring.

In particolare guarda gli ultimi due e osserva come l'estrazione del metodo per la valutazione dei parametri sia propedeutica all'introduzione della lambda. Ne definisce chiaramente input e output. Per me è comunque soddisfacente anche senza lambda, in questo caso. Decidi tu.

L'introduzione della lambda segue un Design Pattern noto come [Strategy](https://refactoring.guru/design-patterns/strategy/java/example#:~:text=Strategy%20is%20a%20behavioral%20design,to%20the%20linked%20strategy%20object.) (anche se con la lambda non sembra, perchè i design pattern sono definiti sull'utilizzo degli oggetti)

Reply:
Chiaro 👍. Grazie mille.
La soluzione con la lambda mi piace particolarmente perché rende il metodo molto più versatile. Che è quello che ho cercato di fare con il metodo che avevo fatto io.

Però ho visto che hai tolto la parte di query `ORDER BY sent_at DESC`.
Gran parte della complessità del mio metodo derivava dal fatto che c'era quello, che andava necessariamente messo prima di `LIMIT ?` altrimenti mi dava l'errore BadSqlGrammar.
Lo avevo messo perché altrimenti il limite imposto ai messaggi era a partire dai primi mandati. Cioè che se gli chiedevo di restituirmi solo 2 messaggi, il db mi mandava i 2 più vecchi (come succede nella tua versione), non i 2 più recenti.
Avevo pensato di metterlo all'interno del metodo generale per prendere i messaggi (quello che avevo chiamato `get` per intenderci) perché mi sembrava sensato che con ogni query i messaggi venissero restituiti con ordine uguale, eliminando la duplicazione derivata dal doverglielo chiedere ogni volta all'invocazione di `get`.

Il fatto che i messaggi vengano ritornati dal più recente al più vecchio è un problema per il front-end?
Perché in caso bisogna trovare un altro modo per fare la query al db, oppure una volta ottenuti in ordine contrario vanno invertiti.

Reply:
Escluso questo problema direi che puoi anche mergiare direttamente il tuo ramo, o mergiarlo in questo per non dover creare un'altra PR, dato che non penso che mi uscirebbe una soluzione migliore.
Tanto le altre modifiche che avevo fatto prima ci sono lo stesso.

Reply:
Hai ragione. Non ho messo `ORDER BY sent_at DESC` perchè l'ho dimenticato. Però aggiungerlo adesso è banale. Lascio fare a te sia l'aggiunta che il merge del mio ramo in questo (buon divertimento 🤣. Scherzo, se lo fai nel modo giusto è facile).